### ISSUE TITLE
Criteria for determining whether speculative_id is accepted

### ISSUE THREAD

**Issue by gaoyonghero** [huggingface/text-generation-inference#1429]
https://github.com/huggingface/text-generation-inference/blob/da27fbdfdbaaab4157e0aa6114c16503471c499c/server/text_generation_server/utils/tokens.py#L290

Whether the next_id and speculative_id are equal is used as a criterion to determine whether to accept it. Will this criterion reduce acceptance rates?

In some papers, the method of comparing the logits of two models was used to determine whether to accept the speculative_id. If directly comparing the next_id with the speculative_id would reduce the acceptance rate, would it have a significant impact on the overall performance of speculative inference?

**Comment by Narsil** [comment]
No this condition is necessary, if you calculated on the wrong speculated ids, then the kv cache is invalid.

There are many variants and things we could have done differently, this is not one of them. Tree speculation is too costly in compute power for vanishing returns. https://github.com/huggingface/api-inference/pull/new/fix_optimum2
